### PR TITLE
Refactor how API routes are defined

### PR DESCRIPTION
**Pull request by guillaumerose**
This improves readability:

``` 
server.POST("/poweroff", handler.PowerOff)
server.GET("/status", handler.Status)
...
```

Handlers can now return an error or JSON.
```
func (h *Handler) Stop(c *context) error {
	_, err := someFunction()
	if err != nil {
		return err
	}
	return c.JSON(http.StatusOK, myStruct{
		Success: true,
	})
}
``` 

Let me know what you think!
I took my inspiration from https://echo.labstack.com/guide/ but I dont really want to vendor it as it is really big lib and we need a small subset finally.

### CODE REVIEW

File: pkg/crc/api/helpers.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+func (s *server) DELETE(pattern string, handler func(c *context) error) {
+	if _, ok := s.routes[pattern]; !ok {
+		s.routes[pattern] = make(map[string]func(*context) error)
+	}
+	s.routes[pattern][http.MethodDelete] = handler
+}
+
+func (s *server) Handler() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		s.routesLock.RLock()
```
Comment (cfergeau) [inline pkg/crc/api/helpers.go:75]: `GET`/`POST`/... should `WLock()` it, or you could remove `routesLock` altogether?

**Review by cfergeau — COMMENTED** [review]
Looks cleaner this way, low level HTTP interactions are nicely wrapped in their own API, which makes the rest of the code easier to read.

File: pkg/crc/api/api_http.go [hunk 2]
```diff
@@ -1,215 +1,66 @@
...
-		} else {
-			w.WriteHeader(http.StatusNotFound)
-		}
-	}
-}
+	server.GET("/webconsoleurl", handler.GetWebconsoleInfo)
 
-func newConfigMux(handler *Handler) http.Handler {
-	mux := http.NewServeMux()
+	server.GET("/config/set", handler.SetConfig)
```
Comment (gbraad) [inline pkg/crc/api/api_http.go:32]: hopefully this makes it easier to refactor these; as these do not follow the verbs #2580

**Review by gbraad — APPROVED** [review]

**Comment by gbraad** [comment]
I love the routing clarity. Reminds me of some of the older Ruby projects I worked on ;-)

File: pkg/crc/api/handlers.go [hunk 3]
```diff
@@ -17,10 +17,10 @@ import (
 )
 
 type Handler struct {
-	Logger        Logger
-	MachineClient AdaptedClient
```
Comment (gbraad) [inline pkg/crc/api/handlers.go:21]: 👍

File: pkg/crc/api/api_http.go [hunk 4]
```diff
@@ -1,215 +1,66 @@
...
+	server := newServer()
 
-	mux.HandleFunc("/status", func(w http.ResponseWriter, r *http.Request) {
-		if wrongHTTPMethodUsed(r, w, http.MethodGet) {
-			return
-		}
-		status := handler.Status()
-		sendResponse(w, status)
-	})
+	server.POST("/start", handler.Start)
```
Comment (gbraad) [inline pkg/crc/api/api_http.go:16]: Like the clarity

**Comment by praveenkumar** [comment]
/hold 

I am still looking some of the changes, will remove hold by EOD.

**Review by gbraad — APPROVED** [review]

**Pull request merged**